Enable ELF support

jtraue opened the pull request:
Support passing ELF binaries as kernel.

So far, the loader expected that a Linux kernel is passed and followed the 32 bit Linux boot protocol. The loader supports now 32 bit ELF binaries as an alternative. If an ELF is provided, the loader follows the Multiboot protocol.

## Changes

* Detect whether an ELF or a Linux kernel is provided.
* Extract the ELF to main memory according to the ELF header.
* Prepare a Multiboot information structure.
* Embed the optional command line in the Multiboot information structure.
* Jump to the ELF's entry point.

jtraue committed:
Verify that a Linux kernel is passed.

The loader will soon get support for booting ELF binaries. This is a
preparation that enables the loader to detect whether a Linux kernel or
an ELF is passed.

jtraue committed:
Use generic naming in firmware config items.

This is a preparation for ELF support. Refering to an ELF binary as
"bzimage" is misleading, so it is replaced by the more generic "kernel".

jtraue committed:
Rename an internal variable.

This is a preparation for ELF support which uses more generic names.

jtraue committed:
Rename an internal data structure.

This is a preparation for ELF support which uses more generic names.

jtraue committed:
Add support for passing ELF binaries as kernel.

Instead of a Linux kernel, a 32 bit ELF binary can now boot as
kernel. The boot follows Multiboot specification and provides Multiboot
information to the kernel. This Multiboot information structure contains
the optional command line.

jtraue committed:
Fix qemu in example in README.

parthy commented on README.md:
```diff
@@ -1,13 +1,18 @@
+The payload assumes that a qemu [fw_cfg](https://github.com/qemu/qemu/blob/master/docs/specs/fw_cfg.txt) device is available and obtains the exact memory addresses from this device. During the boot process, the payload inspects the following items in `fw_cfg`:
+
+- `opt/de.cyberus-technology/kernel_addr`: (mandatory) the start address of the kernel in
+>>>>>>> d5eb3ec (Update README with ELF instructions)
```
This looks like an accidental addition.

parthy commented on elf_boot.h:
```diff
@@ -0,0 +1,61 @@
+    uint32_t flags;     // 32 bit only for now
+    uint16_t ehsize;    // 32 bit only for now
+    uint16_t phentsize; // 32 bit only for now
+    uint16_t phnum;     // 32 bit only for now
```
These repeated comments seem overly noisy. How about naming the structure `elf32_header` and add a bit comment at the top saying that only 32-bit ELF files are supported?

jtraue replied:
Good point. Switched the name. The 32-bit only support is mentioned quite a few times, so I omitted an extra note.

parthy commented on main.c:
```diff
@@ -21,23 +22,72 @@ static void die_on(bool condition, char *string, ...)
+    }
+
+    // ELF binaries begin with a magic number.
+    if (*(uint32_t *)kernel_addr == 0x464c457f) {
```
Should this magic number live in `elf_boot.h`?

parthy replied:
Maybe it makes sense to do this check first, before accessing memory at an offset (`0x202`) that might now contain meaningful data. Worst case, the ELF has the "HdrS" magic at the offset for some obscure reason.

jtraue replied:
You are right - I didn't think of it.

parthy commented on main.c:
```diff
@@ -21,23 +22,72 @@ static void die_on(bool condition, char *string, ...)
+
+    // Linux kernel supporting the "new" boot protocol have a magic number at a specific offset.
+    // See https://www.kernel.org/doc/Documentation/x86/boot.txt.
+    if (*(uint32_t *)(kernel_addr + 0x202) == 0x53726448) {
```
Could this be written using `memcmp((uint32_t *)(kernel_addr + 0x202), "HdrS", 4)`?

blitz replied:
`memcmp((const char *)(...), ...)` would be much better, because it has no strict aliasing issues. Same below.

jtraue replied:
Good point. Fixed.

parthy commented on main.c:
```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
+            elf_addr_in_memory + current_program_header->offset;
+
+        // Pray that nothing is where the ELF contents want to be.
+        // XXX: Is there a way to check it?
```
Maybe the [sysinfo lib](https://github.com/coreboot/coreboot/blob/master/payloads/libpayload/arch/arm/sysinfo.c) helps?

jtraue replied:
What you suggest would at least indicate that the ELF is in usable memory. In addition to using unavailable memory, the target region could also overlap with:
* data provided as payload (the ELF header which is currently parsed or the command line)
* coreboot's heap which is used to allocate multiboot information

I didn't find a documentation regarding the coreboot memory layout and need to dig a bit deeper.

jtraue replied:
The coreboot heap is in a memory region which is reported as "CB_MEM_RAM" by lib_sysinfo, but I can't find more details. We now allocate the multiboot information first and unpack the ELF later. If  ELF unpacking overwrites the multiboot info, the multiboot checksum should give the final program a chance to detect a corruption.

I also added a check for the command line which might get overwritten during ELF unpacking. And ELF unpacking checks that the target region is in usable memory.

That should be the best we can do.

blitz commented on main.c:
```diff
@@ -21,23 +22,72 @@ static void die_on(bool condition, char *string, ...)
+        elf_boot(params);
+        break;
+    default:
+        die_on(true, "Unknown boot method\n");
```
User will see this error if they misconfigure things and this is not actionable. What about "Failed to recognize kernel file format. Supported formats are bzImage and 32-bit Multiboot ELF".

jtraue replied:
That's right. I also converted the other user facing error messages.

blitz commented on main.c:
```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
-    memcpy((char *)boot_params + LINUX_HEADER_OFFSET,
-           (char *)(bzimage_params.bzimage_addr + LINUX_HEADER_OFFSET),
+    memcpy((char *)linux_params + LINUX_HEADER_OFFSET,
+           (char *)(boot_params.kernel_addr + LINUX_HEADER_OFFSET),
```
nit: While you are here, can you make the source operands of `memcpy` `const` pointers?

jtraue replied:
Fixed.

blitz commented on main.c:
```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
+    printf("Loading ELF from address: %p\n", elf);
+
+    bool is_elf_executable_or_unknown =
+        elf->type == ELF_TYPE_EXECUTABLE || elf->type == ELF_TYPE_UNKNOWN;
```
Do we produce binaries with ELF type unknown somewhere? This would be weird.

jtraue replied:
Thanks for pointing that out. It turned out to be an error in the `elf32_header`.

blitz commented on main.c:
```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
+    die_on(!is_elf_executable_or_unknown, "Unknown ELF type 0x%x!\n", elf->type);
+    die_on(elf->class == 2, "64bit ELF not yet supported!\n");
+    die_on(
+        elf->phentsize != 32,
```
`!= sizeof(elf_program_header)` ?

blitz commented on main.c:
```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
+    die_on(elf->class == 2, "64bit ELF not yet supported!\n");
+    die_on(
+        elf->phentsize != 32,
+        "32bit ELF file should have program header entries with size 32, but found size %x!\n",
```
nit: The rest of the code uses `32-bit`.

jtraue replied:
Fixed.

blitz commented on main.c:
```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
+        elf->phentsize);
+
+    // The ELF program header table describes how to prepare the memory layout.
+    printf("ELF program header table offset is 0x%x\n", elf->phoff);
```
nit: I would limit the amout of printfs that the user sees by default. But YMMV.

blitz commented on main.c:
```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
+
+    // The ELF file is loaded to memory already. We need that start addressto copy its contents
+    // (such as code and data) to the correct location.
+    uint8_t *elf_addr_in_memory = (uint8_t *)elf;
```
Why not `const`?

jtraue replied:
Introduced `const` in several more places.

blitz commented on main.c:
```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
+    for (uint16_t i = 0; i < elf->phnum; i++) {
+
+        printf(
+            "Preparing ELF program header (type 0x%x, offset 0x%x, vaddr %p, paddr %p, filesize 0x%x, memsize 0x%x)\n",
```
`Loading ELF segment ...` ?

jtraue replied:
Fixed.

blitz commented on main.c:
```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
+
+    // Prepare Multiboot information.
+    struct mb_boot_information_required *mbinfo = malloc(sizeof(*mbinfo));
+    mbinfo->flags = 0;
```
`memset` with 0 is a better option here to be sure everything is initialized and no random memory leaks into the structure.

jtraue replied:
Fixed.

blitz reviewed (changes requested):
Looks good! I think the remaining feedback should be easy to address. :+1:

tfc commented on main.c:
```diff
@@ -21,23 +22,72 @@ static void die_on(bool condition, char *string, ...)
+// Identify the type of the kernel.
+static enum boot_protocol get_boot_protocol(struct boot_params params)
+{
+    size_t kernel_addr = params.kernel_addr;
```
if this variable intends to store memory addresses as integer numbers, then the type `uintptr_t` is the type that is designed to be used in such cases.

jtraue replied:
Fixed.

jtraue committed:
cleanup: switch to uintptr_t

jtraue committed:
cleanup: use const

blitz commented on elf_boot.h:
```diff
@@ -0,0 +1,66 @@
+struct elf32_program_header {
+    uint32_t type;
+    uint32_t offset;
+    uint8_t *vaddr;
```
I didn't see this before. But this is definitely not a pointer that can/should be used.

For the paddr it's also not really ok, because it will break when we switch to 64-bit payloads or don't have 1:1 mappings of physical memory.

I suggest using uint32_t and casting through uintptr_t to the required pointer type to avoid sign extension.

jtraue replied:
Fixed.

blitz commented on main.c:
```diff
@@ -54,87 +105,178 @@ static struct bzimage_params initialize_from_fw_cfg()
+        // The sysinfo lib could be used to detect whether the destination is usable RAM.
+        // When adding such a check, determine which memory regions coreboot uses
+        // so that upcoming memory allocations don't conflict.
+        const uintptr_t dest_addr = (uintptr_t)current_program_header->vaddr;
```
The ELF needs to be unpacked to what it's paddr says. See for example the output of readelf on hedron for an elf where paddr and  vaddr are not the same.

jtraue replied:
I get that this can be crucial in other scenarios. But we have a 1:1 mapping when this coreboot payload runs. At the end, we directly jump to the ELF's entry point. What difference does using `paddr` and `vaddr` make in that case? Wouldn't it be better to abort if the ELF's `vaddr` and `paddr` differ?

parthy replied:
The issue is not with the loader not running with 1:1 mappings, but the loaded binary. If your code tried to load Hedron, it would attempt to load some segments at negative addresses (`0xffffffff8...`). All the loader needs to do is use the value in `paddr` and the kernel will take care of the rest.

jtraue replied:
Thanks for the (offline) clarification - I turned it into a comment. For reference, ipxe (https://github.com/ipxe/ipxe/blob/master/src/image/elf.c) and grub (https://github.com/rhboot/grub2/blob/ef36b3c93d27bdfc7a68469be33bd8be95630e6a/grub-core/loader/multiboot_elfxx.c) also use `paddr`.

blitz reviewed (changes requested):
I've spotted one more functional problem. Looks good otherwise.

jtraue committed:
Abort if ELF wants to be loaded to address 0

Raise an error if the physical address is set to 0.  Alternatively, the
virtual address could be used if the physical one is unset (as in ipxe
https://github.com/ipxe/ipxe/blob/master/src/image/elf.c#L108).

jtraue committed:
Dump memory map on error.

jtraue committed:
Abort ELF loading on memory overwrites

jtraue committed:
Prepare multiboot information first

jtraue commented:
The loader now checks that ELF unpacking does not overwrite other data. These changes are in separate commits. Most of the other feedback has been updated in the existing commits.

kaldonir commented on main.c:
```diff
@@ -54,87 +124,244 @@ static struct bzimage_params initialize_from_fw_cfg()
+    const struct memory_region higher = first.addr <= second.addr ? second : first;
+
+    // Completely disjunct regions?
+    if (higher.addr > lower.addr + lower.size) {
```
Right now, this condition demands that there is at least one byte space between the regions. Is this intentional?

jtraue replied:
No, thanks for the hint.

kaldonir commented on main.c:
```diff
@@ -54,87 +124,244 @@ static struct bzimage_params initialize_from_fw_cfg()
+            printf("Checking RAM region 0x%lx - -0x%lx\n", coreboot_region.addr,
+                   coreboot_region.addr + coreboot_region.size);
+            if (contains(coreboot_region, region)) {
+                result = true;
```
Can't we simply `return true;` here, and save us the result variable?

jtraue replied:
Yes, that was a left over. It's gone.

kaldonir commented on main.c:
```diff
@@ -54,87 +124,244 @@ static struct bzimage_params initialize_from_fw_cfg()
+
+        // Copy contents of the region from the ELF file.
+        memcpy((void *)elf_region.addr, current_elf_segment_in_memory,
+               current_program_header->filesize);
```
Are we sure that src and dst don't overlap here?

jtraue replied:
Good point. I switched to `memmove` (which takes care of overlapping regions). However, we may still overwrite the ELF which we are currently parsing.

kaldonir commented on main.c:
```diff
@@ -54,87 +124,244 @@ static struct bzimage_params initialize_from_fw_cfg()
+    return true;
+}
+
+bool contains(const struct memory_region container, const struct memory_region containee)
```
Would it make sense to add unit tests for the `contains` and `memory_regions_overlap` functions?

jtraue replied:
In the long term, yes. But it feels like starting on the wrong end here. The major cause for overlapping memory regions is that the payload is already placed in memory when the loader starts. Letting the loader choose the target location would reduce the risks of conflicts. I suggest to address this in the next steps and start with #2 (while keeping #3 in mind).
